### PR TITLE
feat: new leaderboard response with appropriate format

### PR DESCRIPTION
**Pull request by ashw1nr**

### CODE REVIEW

File: src/api/attack/util.rs [hunk 1]
```diff
@@ -190,41 +192,50 @@ pub fn add_game(
...
+    page: i64,
+    limit: i64,
     conn: &mut PgConnection,
-) -> Result<GameHistoryResponse> {
+) -> Result<HistoryboardResponse> {
     use crate::schema::{game, levels_fixture, map_layout};
-    let attacker = fetch_user(conn, attacker_id)?.ok_or(AuthError::UserNotFound)?;
     let joined_table = game::table
         .inner_join(map_layout::table.inner_join(levels_fixture::table))
         .inner_join(user::table.on(game::defend_id.eq(user::id)));
```
Comment (DonWick32) [inline src/api/attack/util.rs]: Remove unnecessary joins if not needed

File: src/api/attack/util.rs [hunk 2]
```diff
@@ -190,41 +192,50 @@ pub fn add_game(
...
+    let last_page: i64 = (total_entries as f64 / limit as f64).ceil() as i64;
+
+    let games_result: Result<Vec<HistoryboardEntry>> = joined_table
+        .filter(game::attack_id.eq(user_id))
+        .offset(off_set)
+        .limit(limit)
         .load::<(Game, (MapLayout, LevelsFixture), User)>(conn)?
         .into_iter()
-        .map(|(game, (_, levels_fixture), defender)| {
+        .map(|(game, (_, levels_fixture), _defender)| {
```
Comment (DonWick32) [inline src/api/attack/util.rs]: Remove unused params

File: src/api/attack/util.rs [hunk 3]
```diff
@@ -190,41 +192,50 @@ pub fn add_game(
...
-                    username: defender.username,
-                    trophies: defender.trophies,
-                    avatar_id: defender.avatar_id,
-                },
-                is_replay_available,
+            Ok(HistoryboardEntry {
+                opponent_user_id: game.defend_id,
+                is_attack: true,
+                damage_percent: game.damage_done,
+                result: "none".to_string(),
```
Comment (DonWick32) [inline src/api/attack/util.rs]: What is this result?

File: src/api/defense/util.rs [hunk 4]
```diff
@@ -379,47 +380,55 @@ pub fn set_map_invalid(conn: &mut PgConnection, map_id: i32) -> Result<()> {
...
+pub fn fetch_defense_historyboard(
     user_id: i32,
+    page: i64,
+    limit: i64,
     conn: &mut PgConnection,
-) -> Result<GameHistoryResponse> {
+) -> Result<HistoryboardResponse> {
     use crate::schema::{game, levels_fixture, map_layout};
 
     let joined_table = game::table.inner_join(map_layout::table.inner_join(levels_fixture::table));
```
Comment (DonWick32) [inline src/api/defense/util.rs]: Here too, don't join if not required

File: src/api/defense/util.rs [hunk 5]
```diff
@@ -379,47 +380,55 @@ pub fn set_map_invalid(conn: &mut PgConnection, map_id: i32) -> Result<()> {
...
-                    trophies: defender.trophies,
-                    avatar_id: defender.avatar_id,
-                },
-                is_replay_available,
+            Ok(HistoryboardEntry {
+                opponent_user_id: game.attack_id,
+                is_attack: false,
+                damage_percent: game.damage_done,
+                result: "none".to_string(),
+                artifacts_taken: game.artifacts_collected,
```
Comment (DonWick32) [inline src/api/defense/util.rs]: Shouldn't this be -ve, cuz it is the number of artifacts lost

File: src/api/defense/util.rs [hunk 6]
```diff
@@ -379,47 +380,55 @@ pub fn set_map_invalid(conn: &mut PgConnection, map_id: i32) -> Result<()> {
...
+        .limit(limit)
         .load::<(Game, (MapLayout, LevelsFixture))>(conn)
         .map_err(|err| DieselError {
             table: "game",
             function: function!(),
             error: err,
         })?
         .into_iter()
-        .map(|(game, (_, levels_fixture))| {
+        .map(|(game, (_map_layout, levels_fixture))| {
```
Comment (DonWick32) [inline src/api/defense/util.rs]: Remove extra params if not used

File: src/api/game/util.rs [hunk 7]
```diff
@@ -24,76 +28,24 @@ pub struct LeaderboardResponse {
...
-
-#[derive(Queryable, Deserialize, Serialize)]
-pub struct LeaderboardEntry {
-    pub attacker: UserDetail,
-    pub defender: UserDetail,
-    pub can_be_attacked: bool,
+    pub artifacts: i32,
+    pub attacks_won: i32,
+    pub defenses_won: i32,
+    pub avatar_url: i32,
```
Comment (DonWick32) [inline src/api/game/util.rs:38]: change it to avatar_id

File: src/api/game/util.rs [hunk 8]
```diff
@@ -24,76 +28,24 @@ pub struct LeaderboardResponse {
...
 }
 
 pub fn get_leaderboard(
     page: i64,
     limit: i64,
-    user_id: i32,
     conn: &mut PgConnection,
 ) -> Result<LeaderboardResponse> {
-    use crate::schema::{game, map_layout, user};
+    use crate::schema::{map_layout, user};
```
Comment (DonWick32) [inline src/api/game/util.rs]: Is map_layout required?

**Review by DonWick32 — CHANGES_REQUESTED** [review]

File: src/api/attack/util.rs (hunk 2 above)
Comment (ashw1nr) [reply src/api/attack/util.rs]: Okay

File: src/api/attack/util.rs (hunk 1 above)
Comment (ashw1nr) [reply src/api/attack/util.rs]: We can reach the levels_fixture table, which is used while checking for replay availability, from game table through map_layout table

File: src/api/defense/util.rs (hunk 5 above)
Comment (ashw1nr) [reply src/api/defense/util.rs]: You are right!

**Review by DonWick32 — APPROVED** [review]

**Pull request merged**